internal/scanner/layerscanner/layerscanner.go

```diff
@@ -0,0 +1,175 @@
+package layerscanner
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/internal/scanner"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rs/zerolog"
+)
+
+// layerScanner implements the scanner.LayerScanner interface.
+type layerScanner struct {
+	// common depedencies
+	*scanner.Opts
+	// concurrency level. maximum number of concurrent layer scans
+	cLevel int
+	// a channel to implement concurrency control
+	cc chan struct{}
+	// a logger that can be called with context
+	logger zerolog.Logger
+}
+
+// New is a constructor for a defaultLayerScanner
+func New(cLevel int, opts *scanner.Opts) scanner.LayerScanner {
+	return &layerScanner{
+		Opts:   opts,
+		cLevel: cLevel,
+	}
+}
+
+// addToken will block until a spot in the conccurency channel is available
+// or the ctx is canceled.
+func (ls *layerScanner) addToken(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ls.cc <- struct{}{}:
+		return nil
+	}
+}
+
+// discardToken is only called after addToken. Removes a token
+// from the concurrency channel allowing another task to kick off.
+func (ls *layerScanner) discardToken() {
+	select {
+	case <-ls.cc:
+	default:
+	}
+}
+
+// Scan performs a concurrency controlled scan of each layer by each type of configured scanner; indexing
+// the results on successful completion.
+//
+// Scan will launch all necessary go routines and each routine will block on adding a token.
+// On completion a token is discarded unblocking other routines which are waiting.
+//
+// On ctx cancel or a go routine reporting an scan/index error all routines blocking on adding a token will error
+// and the will not subsequently try to discard a token.
```

Comment:
Again, what are tokens? This should probably start "On cancellation or the first reported non-nil error,". There's also something missing here.